python-package/lightgbm/callback.py

```diff
@@ -209,12 +215,69 @@ def _init(env):
                 best_score.append(float('inf'))
                 cmp_op.append(lt)
 
+    def _metric_alias_matching(target_metric):
+        """Convert metric name for early stopping from alias to representative in order to match the param metric."""
+        metrics_list = [("l1", {"mean_absolute_error", "mae", "regression_l1"}),
+                        ("l2", {"mean_squared_error", "mse", "regression_l2", "regression"}),
+                        ("l2_root", {"root_mean_squared_error", "rmse"}),
+                        ("mape", {"mean_absolute_percentage_error"}),
+                        ("ndcg", {"lambdarank"}),
+                        ("map", {"mean_average_precision"}),
+                        ("binary_logloss", {"binary"}),
+                        ("multi_logloss",
+                        {"multiclass", "softmax", "multiclassova", "multiclass_ova", "ova", "ovr"}),
+                        ("xentropy", {"cross_entropy"}),
+                        ("xentlambda", {"cross_entropy_lambda"}),
+                        ("kldiv", {"kullback_leibler"})]
+        for metric, aliases in metrics_list:
+            if target_metric in aliases:
+                return metric
+        return target_metric
+
     def _callback(env):
         if not cmp_op:
             _init(env)
         if not enabled[0]:
             return
+        if first_metric_only:
+            # Choosing a target metric for early stopping from first element of metrics list as `eval_metric`
+            # if first_metric_only == True
+            eval_metric = None
+            for metric_alias in ['metric', 'metrics', 'metric_types']:
+                if metric_alias in env.params.keys():
+                    if isinstance(env.params[metric_alias], (tuple, list)):
+                        metric_list = [m for m in env.params[metric_alias] if m is not None]
+                        eval_metric = metric_list[0]
```

Comment:
@matsuken92 During reading this part of source code, I imagined the following (quite strange, but still valid) edge case for `metric` param:
```
from sklearn.datasets import load_boston
from sklearn.model_selection import train_test_split

import lightgbm as lgb

X_train, X_test, y_train, y_test = train_test_split(*load_boston(True), test_size=0.2, random_state=42)
lgb_train = lgb.Dataset(X_train, y_train)
lgb_valid = lgb.Dataset(X_test, y_test, reference=lgb_train)
lgb.train({'objective': 'regression', 'metric': [], 'learning_rate': 1, 'first_metric_only': True},
          lgb_train, num_boost_round=30, valid_sets=[lgb_valid], early_stopping_rounds=1)
```
`metric=[]` should be treated in the same way as the default one `metric=None`. Set `first_metric_only=False` in my example above to observe this.
Honestly, I imagined the case with `metric=[None]`, but this case fails earlier during `_init` phase of callback:
```
ValueError: For early stopping, at least one dataset and eval metric is required for evaluation
```

Reply:
I guess something like this should help:
```
eval_metric = metric_list[0] if len(metric_list) > 0 else None
```